### PR TITLE
Comic Book Archives

### LINKED ISSUE #3: Add CBZ/CBR support.
I'm not yet sure how would the user switch between comic books and images within a single cb but this may be fairly straingthforward to implement and shouldn't be intrusive for those who don't use this feature.

### PR DESCRIPTION
**Pull request by Aloso**
Sorry for being inactive for so long. This is my new attempt to implement support for comic book archives. Like the previous attempt, it depends on `libarchive` to be installed on the system when compiling emulsion. Therefore, it is feature-gated behind the `archives` feature.

Closes #3

The default key bindings for turning the page are <kbd>ArrowUp</kbd> and <kbd>ArrowDown</kbd>.

Within comic book archives, only image types without animations that are supported by the `image` crate are accepted. This is okay, since niche image formats such as AVIF or animated images are never used in comic book archives, AFAIK.

@ArturKovacs this is ready for review!

### CODE REVIEW

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -87,4 +88,4 @@ jobs:
         shell: cmd
         run: |
           call "C:/Program Files (x86)/Microsoft Visual Studio/2019/Enterprise/VC/Auxiliary/Build/vcvars64.bat"
-          env RUSTFLAGS="-C opt-level=0" cargo test --verbose --all-features
+          env RUSTFLAGS="-C opt-level=0" cargo test --verbose --features=avif,networking
```
Comment (Aloso) [inline .github/workflows/test.yml:91]: This disables the `archives` feature, so that the build doesn't fail

File: src/playback_manager.rs [hunk 2]
```diff
@@ -342,14 +340,9 @@ impl<P: Playback> ImgSequencePlayer<P> {
...
-		match self.playback_state {
-			PlaybackState::Present | PlaybackState::RandomPresent => {
-				frame_delta_time_nanos = (NANOS_PER_SEC * 6) as i64;
-			}
-			_ => {
-				frame_delta_time_nanos = P::delay_nanos(&self) as i64;
-			}
+		let frame_delta_time_nanos = match self.playback_state {
+			PlaybackState::Present | PlaybackState::RandomPresent => (NANOS_PER_SEC * 6) as i64,
+			_ => P::delay_nanos(&self) as i64,
```
Comment (Aloso) [inline src/playback_manager.rs:345]: I applied some `clippy` suggestions here

**Comment by ArturKovacs** [comment]
I'm currently trying to get this to compile on my Windows machine.

I previously installed `libarchive` with the `x64-windows-static-md` triplet but during the compilation of Emulsion I got LNK2019 errors. I then installed `libarchive` with `x64-windows-static` and tried again, but it seems that I'm getting the same error at the same stage.
```
error: linking with `link.exe` failed: exit code: 1120
  |
  = note: "C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\Community\\VC\\Tools\\MSVC\\14.28.29333\\bin\\HostX64\\x64\\link.exe" "/NOLOGO" "/NXCOMPAT" "/LIBPATH:C:\\Users\\kovac\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib" "D:\\personal\\software\\contribution\\emulsion\\target\\release\\deps\\emulsion.emulsion.9i8uunir-cgu.5.rcgu.o" "/OUT:D:\\personal\\software\\contribution\\emulsion\\target\\release\\deps\\emulsion.exe" "/SUBSYSTEM:windows" "/ENTRY:mainCRTStartup" "/OPT:REF,ICF" "/DEBUG" "/NATVIS:C:\\Users\\kovac\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\etc\\intrinsic.natvis" "/NATVIS:C:\\Users\\kovac\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\etc\\liballoc.natvis" "/NATVIS:C:\\Users\\kovac\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\etc\\libcore.natvis" "/NATVIS:C:\\Users\\kovac\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\etc\\libstd.natvis" "/LIBPATH:D:\\personal\\software\\contribution\\emulsion\\target\\release\\deps" "/LIBPATH:D:\\personal\\software\\contribution\\emulsion\\target\\release\\build\\emulsion-2ba3653cbb241ae5\\out" "/LIBPATH:D:\\personal\\software\\third-party\\vcpkg-2020-11-1\\installed\\x64-windows-static\\lib" "/LIBPATH:D:\\personal\\software\\contribution\\emulsion\\target\\release\\build\\sys-info-2e09701aa22b9030\\out" "/LIBPATH:C:\\Users\\kovac\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib" "resource.lib" "C:\\Users\\kovac\\AppData\\Local\\Temp\\rustcjr3LZg\\libsys_info-5b81334254ad9169.rlib" "C:\\Users\\kovac\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\libcompiler_builtins-937aede5bfe8081e.rlib" "psapi.lib" "archive.lib" "zstd.lib" "libcrypto.lib" "libssl.lib" "lzo2.lib" "lzma.lib" "lz4.lib" "xxhash.lib" "libxml2.lib" "zlib.lib" "libcharset.lib" "libiconv.lib" "bz2.lib" "advapi32.lib" "opengl32.lib" "advapi32.lib" "cfgmgr32.lib" "comctl32.lib" "credui.lib" "dwmapi.lib" "fwpuclnt.lib" "gdi32.lib" "kernel32.lib" "msimg32.lib" "ntdll.lib" "ole32.lib" "opengl32.lib" "secur32.lib" "shcore.lib" "shell32.lib" "shlwapi.lib" "synchronization.lib" "user32.lib" "uxtheme.lib" "winspool.lib" "ws2_32.lib" "advapi32.lib" "ws2_32.lib" "userenv.lib" "libcmt.lib"
  = note: libcrypto.lib(e_capi.obj) : error LNK2019: unresolved external symbol __imp_CertOpenStore referenced in function capi_list_certs
          libcrypto.lib(e_capi.obj) : error LNK2019: unresolved external symbol __imp_CertCloseStore referenced in function capi_find_key
          libcrypto.lib(e_capi.obj) : error LNK2019: unresolved external symbol __imp_CertEnumCertificatesInStore referenced in function capi_find_cert
          libcrypto.lib(e_capi.obj) : error LNK2019: unresolved external symbol __imp_CertFindCertificateInStore referenced in function capi_find_cert
          libcrypto.lib(e_capi.obj) : error LNK2019: unresolved external symbol __imp_CertDuplicateCertificateContext referenced in function capi_load_ssl_client_cert
          libcrypto.lib(e_capi.obj) : error LNK2019: unresolved external symbol __imp_CertFreeCertificateContext referenced in function capi_dsa_free
          libcrypto.lib(e_capi.obj) : error LNK2019: unresolved external symbol __imp_CertGetCertificateContextProperty referenced in function capi_cert_get_fname
          D:\personal\software\contribution\emulsion\target\release\deps\emulsion.exe : fatal error LNK1120: 7 unresolved externals

```

It might be related to the fact that I had Visual Studio Build tools 2017 and Visual Studio 2019 both installed.

I removed build tools, built a fresh vcpkg, installed a fresh `libarchive`, ran `cargo clean`, set the `OPENSSL_ROOT_DIR` env var, and ran a new build but I'm still getting the same error.

I'll just test this on Linux; it doesn't seem wortwhile to get this to compile on my Windows. I'm of course gonna have to test the release coming from the CI before publishing the next release. Let's hope I'll remember that :D

**Review by ArturKovacs — CHANGES_REQUESTED** [review]
I tested it on Linux.

First of all, handling the comic book pages like animation frames is a very good approach. However I still believe it would be better to require the user to "enter" the comic book with some key input, and then they could use the regular next/prev keybindings for stepping between pages. 

This could be implemented fairly straightforwardly within `PlaybackManager` (in `playback_manager.rs`). This PR wouldn't have to introduce any fundamental changes to how image caching is done, you would just have to introduce a field in `PlaybackManager` that keeps track whether we are in "file step mode" or "frame step mode" (you can call it whatever you see fit :D).

For example within "frame step mode": `current_dir_len` would return the number of frames within the current file, `request_load` would load the next frame when it receives a `LoadNext`, and so on... This would mean that there would be no need for `LoadRequest::LoadNextFrame` and `LoadRequest::LoadPreviousFrame`

**Comment by Aloso** [comment]
It's very unfortunate that building compress-tools is so difficult. I looked into alternative crates again and submitted an issue on `unrar` (see above). If this crate gets the functionality we need, we can ditch compress-tools. `unrar` depends on a much smaller C++ library that is built automatically during compilation, so it should Just Work, while also being faster to compile.

**Comment by ArturKovacs** [comment]
To be honest I'm not even sure if supporting any comic book format is going to be benefitial for Emulsion. I tried reading a comic book archive using Emulsion and using the default viewer for Pop!_OS, (I think it's Evince). The user experience was waay better with the default viewer. The issue with Emulsion was that the pages were so large that I had to zoom in to read the text. But then I had "pan" down to go from the top of the page to the bottom. But while I was panning I accidentally kept moving the page left and right so I had to invest extra focus to keep the page in the middle. But then when I reached the bottom of the page and flipped to the next page, I had to "pan" back to the top again. Whereas in the default viewer it was showing like a pdf document so I could just keep going down in one continous scroll to go through all the pages. Of course we could change the default behaviour of Emulsion when the user is inside a comic book, but I feel like that would be a bad path to go down.

I know that it's annoying that I only express this now that you've spent so much time on it, but it's only at this point where seriously start to question whether adding this feature will be any good for anyone.

**Comment by ArturKovacs** [comment]
@Aloso I want to be clear on something. The particular implementation you made has nothing to do with the fact that I don't find this feature suited for Emulsion. I definitely *don't* want to deter you from contributing.

**Comment by Aloso** [comment]
That's all right, and I mostly agree with your assessment.

I'm disappointed because I spent a lot of time implementing this, only to learn that emulsion is unsuited for comic book archives. But things like this happen, so let's cut our losses and move on :)

**Pull request closed without merging**